### ISSUE TITLE
[RFC] Support for multiple training algorithms

### ISSUE THREAD

**Issue by SplitInfinity** [pytorch/glow#2554]
**Preface**
In #2524, I added support for the Adagrad training algorithm in preparation for adding training support for `SparseLengthsWeightedSum` in #2541. The design I chose was not well-received, so I want to propose a new design based on the feedback in that PR and discuss it before I update the PR.

**Problem Statement**
`glow::differentiate` adds `SGDNodes` for every trainable `Placeholder` in a Graph that calculate the new updated values based on the stochastic gradient descent algorithm. It also adds `SaveNodes` to save these updated values back into their corresponding `Placeholders`. However, this approach does not allow Glow to support multiple training algorithms (i.e. multiple ways to compute new weights).

**Proposal**
There is a simple POD struct named `TrainingConfig` that includes a bunch of parameters used by stochastic gradient descent:

```
struct TrainingConfig {
  float L1Decay{0};
  float L2Decay{0};
  float learningRate{0.01f};
  float momentum{0.0};
  unsigned batchSize{1};
};
```

This struct is passed into `glow::differentiate` and used to construct `SGDNodes` with the proper learning rate, momentum, etc. I propose that we replace this with the following:

```
class Trainer {
public:
  Trainer(float learningRate, unsigned batchSize) : learningRate_(learningRate), batchSize_(batchSize) {}
  virtual bool instrument(Function* F, PlaceholderBindings& bindings) = 0;
protected:
  // Parameters used by multiple subclasses
  float learningRate_{0.01f};
  unsigned batchSize_{1};
};
```
 
New training algorithms are implemented as follows:

```
// Stochastic gradient descent
class SGDTrainer : final Trainer {
public:
  SGDTrainer(float learningRate, unsigned batchSize, float L1Decay, float L2Decay, float momentum) : Trainer(learningRate, batchSize), L1Decay_(L1Decay), L2Decay_(L2Decay), momentum_(momentum) {}
  bool instrument(Function* F, PlaceholderBindings& bindings) override {
    // code here to add SGDNode to F, save new weights into Placeholders, etc.

   // return true if instrumentation was successful
  }
private:
  float L1Decay{0};
  float L2Decay{0};
  float momentum{0.0};
};

// Adagrad
class AdagradTrainer : final Trainer {
public:
  AdagradTrainer(float learningRate, unsigned batchSize, float epsilon) : Trainer(learningRate, batchSize), epsilon_(epsilon) {}
  bool instrument(Function* F, PlaceholderBindings& bindings) override {
    // code here to add Adagrad to F, create new Placeholders to store momentum history, etc.

   // return true if instrumentation was successful
  }
private:
  float epsilon_{1e-5};
};
```

`glow::differentiate` would accept a reference/pointer to `Trainer` and use it as follows:
```
// Clone the function to produce the training version of it.
auto *G = F->clone(newFunctionName);
...
// Insert grad nodes.
...
// Add training instrumentation.
bool ok = trainer->instrument(G, bindings);

if (!ok) {
  // report error, assert, etc.
}
```

**Open Questions**
1. Should there be getters/setters for the algorithm parameters of `Trainer` and its subclasses? There is some code that currently reads them out of a `TrainingConfig` object, but that's easy enough to modify so that there are other variable in the same scope as `TrainingConfig` that store those values.
